keps/sig-apps/4368-support-managed-by-label-for-batch-jobs/README.md

```diff
@@ -0,0 +1,985 @@
+restarting the component. No other changes should be necessary to disable the
+feature.
+
+NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
+-->
+
+###### What happens if we reenable the feature if it was previously rolled back?
+
+The feature behaves as if it was enabled for the first time.
+
+###### Are there any tests for feature enablement/disablement?
+
+<!--
+The e2e framework does not currently support enabling or disabling feature
+gates. However, unit tests in each component dealing with managing data, created
+with and without the feature, are necessary. At the very least, think about
+conversion tests if API types are being modified.
+
+Additionally, for features that are introducing a new API field, unit tests that
+are exercising the `switch` of feature gate itself (what happens if I disable a
+feature gate after having objects written with the new field) are also critical.
+You can take a look at one potential example of such test in:
+https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
+-->
+
+Given that this feature doesn't introduce any new API field, enablement/disablement tests will not provide reasonable value and won't be added.
+
+### Rollout, Upgrade and Rollback Planning
+
+<!--
+This section must be completed when targeting beta to a release.
+-->
+
+###### How can a rollout or rollback fail? Can it impact already running workloads?
+
+<!--
+Try to be as paranoid as possible - e.g., what if some components will restart
+mid-rollout?
+
+Be sure to consider highly-available clusters, where, for example,
+feature flags will be enabled on some API servers and not others during the
+rollout. Similarly, consider large clusters and how enablement/disablement
+will rollout across nodes.
+-->
+
+The rollout will not impact already running workloads, unless they set the
+`managed-by` label to a custom value, but this would require a prior intentional
+action.
+
+###### What specific metrics should inform a rollback?
+
+A substantial increase in the `apiserver_request_total[code=409, resource=job, group=batch]`,
+while there are jobs with the custom `managed-by` label, can be indicative of
+the built-in job controller stepping onto another controller, causing conflicts.
+This can be further investigate per-job by checking the `.metadata.managedFields.manager`
+being flipped between two owners.
+
+The feature is opt-in so in case of such problems the custom `managed-by` label
+should not be used.
+
+A substantial drop in the `job_sync_duration_seconds`, while the number of
+jobs with the custom `managed-by` label is low, could be indicative of the
+Job controller skipping reconciliation of jobs it should reconcile. This could
+be further investigated per-job by looking at the timestamp of changes in
+`.metadata.managedFields.time`, and owners in `.metadata.managedFields.manager`.
+
```

Comment:
If this feature has an impact on the cronjob controller, it might be worth having metrics on that too.

Reply:
I was actually considering adding a separate metric for external controllers.

Reply:
I think a new metric to capture external controllers is a good idea, but I would do it at the Job level, rather than CronJob. The reason is that when using CronJob, the job template may not have the `managed-by` label, but the label is later added by the MultiKueue webhook, so the metric could be even misleading.

So maybe:
- `by_external_controller_total` counter with label `name` (corresponding to the value of `managed-by`), not counted when the feature is disabled or the value points to the built-in controller.

WDYT? I will update the KEP accordingly.

Reply:
IIUC, that metric sounds useful

For cronjob controller, we could also observe [kube_cronjob_status_active](https://github.com/kubernetes/kube-state-metrics/blob/90bd24f27c618ce916db4cd9da34d854d49db5b6/internal/store/cronjob.go#L132) metric and measure an increase in the active jobs value for a known set of jobs. Or measure an increase in the amount of time the jobs should be running. Either of these would indicate that some jobs are not progressing or syncing the status incorrectly. Similar thing can also be done one level lower for the jobs.

Reply:
If this were for just a job it would probably only be a Troubleshooting since we are not responsible for the external controllers. But a CronJob depends on these jobs and they can affect its health, so it probably belongs here IMO.

Reply:
I see, added a paragraph about this case. PTAL

Reply:
thx

Reply:
> * `job_by_external_controller_total` counter with label `name`

:+1: